pydantic/_internal/_namespace_utils.py

```diff
@@ -0,0 +1,237 @@
+from __future__ import annotations
+
+import sys
+from collections.abc import Generator, Mapping
+from contextlib import contextmanager
+from functools import cached_property
+from typing import Any, Iterator, NamedTuple, TypeAlias, TypeVar
+
+GlobalsNamespace: TypeAlias = dict[str, Any]
+"""A global namespace.
+
+In most cases, this is a reference to the `__dict__` attribute of a module.
+This namespace type is expected as the `globals` argument during annotations evaluation.
+"""
+
+MappingNamespace: TypeAlias = Mapping[str, Any]
+"""Any kind of namespace.
+
+In most cases, this is a local namespace (e.g. the `__dict__` attribute of a class,
+the [`f_locals`][frame.f_locals] attribute of a frame object).
+This namespace type is expected as the `locals` argument during annotations evaluation.
+"""
+
+
+class NamespacesTuple(NamedTuple):
+    """A tuple of globals and locals to be used during annotations evaluation.
+
+    This datastructure is defined as a named tuple so that it can easily be unpacked:
+
+    ```python lint="skip" test="skip"
+    def eval_type(typ: type[Any], ns: NamespacesTuple) -> None:
+        return eval(typ, *ns)
+    ```
+    """
+
+    globals: GlobalsNamespace
+    """The namespace to be used as the `globals` argument during annotations evaluation."""
+
+    locals: MappingNamespace
+    """The namespace to be used as the `locals` argument during annotations evaluation."""
+
+
+def get_module_ns_of(obj: Any) -> dict[str, Any]:
+    """Get the namespace of the module where the object is defined.
+
+    Caution: this function does not return a copy of the module namespace, so it should not be mutated.
+    The burden of enforcing this is on the caller.
+    """
+    module_name = getattr(obj, '__module__', None)
+    if module_name:
+        try:
+            return sys.modules[module_name].__dict__
+        except KeyError:
+            # happens occasionally, see https://github.com/pydantic/pydantic/issues/2363
+            return {}
+    return {}
+
+
+class LazyLocalNamespace(Mapping[str, Any]):
+    """A lazily evaluated mapping, to be used as the `locals` argument during annotations evaluation.
+
+    While the [`eval`][eval] function expects a mapping as the `locals` argument, it only
+    performs `__getitem__` calls. The [`Mapping`][collections.abc.Mapping] abstract base class
+    is fully implemented only for type checking purposes.
+
+    Args:
+        *namespaces: The namespaces to consider, in ascending order of priority.
+
+    Example:
+        ```python lint="skip" test="skip"
+        ns = LazyLocalNamespace({'a': 1, 'b': 2}, {'a': 3})
+        ns['a']
+        #> 3
+        ns['b']
+        #> 2
+        ```
+    """
+
+    def __init__(self, *namespaces: MappingNamespace) -> None:
+        self._namespaces = namespaces
+
+    @cached_property
```

Comment:
Hows the perf looking with this in ForwardRefs cases like in the mega kube file thing when using "__future__" annotations? So how this works by flatting the dicts vs just iterating them? I guess the `def data` is really never active when ForwardRefs are not being used by user right?

Reply:
> I guess the `def data` is really never active when ForwardRefs are not being used by user, right?

Correct.

Running memray on the k8s file, I get:
- ~3 GiB total allocations on main
- ~1.5GiB total allocations here
- ~1.9GiB total allocations here, with the future import.

 > So how this works by flatting the dicts vs just iterating them, perf wise?

Considering most of the local namespaces are really small (compared to globals), If I remember correctly the difference is marginal, at least by experimenting quickly with `timeit` and some dummy examples.

Reply:
This is great, I'm super glad to see we've reduced the duplicate allocations in context management that we were dealing with before this PR.